### PR TITLE
[MIGR]: stock_picking_wave: Migrated into new api.

### PR DESCRIPTION
**Pull request by rgo-odoo**
Task: https://www.odoo.com/web?#id=13157&view_type=form&model=project.task&menu_id=3942&action=327
Pad: https://pad.odoo.com/p/r.yzlRPJQy5dn23RFu

### CODE REVIEW

File: addons/stock_picking_wave/controllers/main.py [hunk 1]
```diff
@@ -1,16 +1,15 @@
-from openerp.addons.web import http
-from openerp.addons.web.http import request
+from openerp import http
+from openerp.http import request
 
 
 class picking_wave_report(http.Controller):
     @http.route('/report/stock_picking_wave.report_pickingwave/<ids>', type='http', auth='user',
```
Comment (rgo-odoo) [inline addons/stock_picking_wave/controllers/main.py]: Note: It seems there isn't any specific option available in official modules to call this controller, this controller may be removed while merging.

File: addons/stock_picking_wave/__openerp__.py [hunk 2]
```diff
@@ -13,13 +13,13 @@
     'website': 'https://www.odoo.com/page/warehouse',
     'depends': ['stock'],
     'data': ['security/ir.model.access.csv',
-            'stock_picking_wave_view.xml',
-            'stock_picking_wave_data.xml',
-            'stock_picking_wave_sequence.xml',
+            'views/stock_picking_wave_view.xml',
```
Comment (rgo-odoo) [inline addons/stock_picking_wave/__openerp__.py]: file should be named as stock_picking_wave_views.xml(do the same change for wizard/picking_to_wave_view.xml)

File: addons/stock_picking_wave/__openerp__.py [hunk 3]
```diff
@@ -13,13 +13,13 @@
...
+            'data/stock_picking_wave_data.xml',
+            'data/stock_picking_wave_sequence.xml',
             'wizard/picking_to_wave_view.xml',
             ],
     'demo': [
-            'stock_picking_wave_demo.xml',
+            'data/stock_picking_wave_demo.xml',
              ],
     'installable': True,
     'auto_install': False,
```
Comment (rgo-odoo) [inline addons/stock_picking_wave/__openerp__.py]: not needed

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# -*- coding: utf-8 -*-
+from openerp import fields, api, models
+from openerp.tools.translate import _
```
Comment (rgo-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py]: from openerp import api, fields, models, _

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+    _description = "Picking Wave"
+    _order = "name desc"
+    name = fields.Char('Picking Wave Name', default='/', required=True, help='Name of the picking wave', copy=False)
+    user_id = fields.Many2one('res.users', 'Responsible', track_visibility='onchange', help='Person responsible for this wave')
+    picking_ids = fields.One2many('stock.picking', 'wave_id', 'Pickings', help='List of picking associated to this wave')
+    state = fields.Selection([('draft', 'Draft'), ('in_progress', 'Running'), ('done', 'Done'), ('cancel', 'Cancelled')], string="State", default ='draft', track_visibility='onchange', required=True, copy=False)
+
+    @api.multi
+    def confirm_picking(self):
+        picking = self.env['stock.picking'].search([('wave_id', 'in', self.ids)])
```
Comment (rgo-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py]: picking --> pickings

File: addons/stock_picking_wave/wizard/picking_to_wave_view.xml [hunk 6]
```diff
@@ -1,7 +1,5 @@
 <?xml version="1.0"?>
```
Comment (rgo-odoo) [inline addons/stock_picking_wave/wizard/picking_to_wave_view.xml]: you missed encoding="utf-8"

File: addons/stock_picking_wave/wizard/picking_to_wave.py [hunk 7]
```diff
@@ -1,16 +1,14 @@
...
+    wave_id = fields.Many2one('stock.picking.wave', 'Picking Wave', required=True)
+
+    def attach_pickings(self):
         #use active_ids to add picking line to the selected wave
-        wave_id = self.browse(cr, uid, ids, context=context)[0].wave_id.id
-        picking_ids = context.get('active_ids', False)
-        return self.pool.get('stock.picking').write(cr, uid, picking_ids, {'wave_id': wave_id})
\ No newline at end of file
+        wave_id = self.wave_id.id
+        return self.env['stock.picking'].write({'wave_id': wave_id})
```
Comment (rgo-odoo) [inline addons/stock_picking_wave/wizard/picking_to_wave.py]: decorate this method with @api.multi and change it's code to,

self.ensure_one()
picking_ids = self.env.context.get('active_ids')
return self.env['stock.picking'].browse(picking_ids).write({'wave_id': self.wave_id.id})

File: addons/stock_picking_wave/__openerp__.py [hunk 8]
```diff
@@ -12,13 +12,13 @@
     'website': 'https://www.odoo.com/page/warehouse',
     'depends': ['stock'],
     'data': ['security/ir.model.access.csv',
-            'stock_picking_wave_view.xml',
-            'stock_picking_wave_data.xml',
-            'stock_picking_wave_sequence.xml',
-            'wizard/picking_to_wave_view.xml',
+            'views/stock_picking_wave_views.xml',
+            'data/stock_picking_wave_data.xml',
+            'data/stock_picking_wave_sequence.xml',
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/__openerp__.py]: By the way, rename this one ir_sequence_data_picking_wave.xml

File: addons/stock_picking_wave/controllers/main.py [hunk 9]
```diff
@@ -1,16 +1,17 @@
...
+from openerp import http
+from openerp.http import request
 
-class picking_wave_report(http.Controller):
-    @http.route('/report/stock_picking_wave.report_pickingwave/<ids>', type='http', auth='user', 
-                website=True)
-    def report_picking_wave(self, ids):
-        self.cr, self.uid, self.pool = request.cr, request.uid, request.registry
+
+class PickingWaveReport(http.Controller):
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/controllers/main.py]: Well, move the removal commit before migrating, so that we don't have to review the migration of something we want to remove ;) .

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 10]
```diff
@@ -1,77 +1,66 @@
...
+    user_id = fields.Many2one('res.users', string='Responsible', track_visibility='onchange', help='Person responsible for this wave')
+    picking_ids = fields.One2many('stock.picking', 'wave_id', string='Pickings', help='List of picking associated to this wave')
+    state = fields.Selection([('draft', 'Draft'), ('in_progress', 'Running'), ('done', 'Done'), ('cancel', 'Cancelled')], default ='draft', track_visibility='onchange', required=True, copy=False)
 
-    def confirm_picking(self, cr, uid, ids, context=None):
-        picking_todo = self.pool.get('stock.picking').search(cr, uid, [('wave_id', 'in', ids)], context=context)
-        self.write(cr, uid, ids, {'state': 'in_progress'}, context=context)
-        return self.pool.get('stock.picking').action_assign(cr, uid, picking_todo, context=context)
+    @api.multi
+    def confirm_picking(self):
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py]: This is potentially a big mistake. You changed the operation order as well as method return, which can lead to regressions. The original method find pickings, update the waves, assign the pickings and return the assignation result. You do not respect that. Please update the method.

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 11]
```diff
@@ -1,77 +1,66 @@
...
-                message_body = "<b>%s:</b> %s <a href=#id=%s&view_type=form&model=stock.picking.wave>%s</a>" % (_("Transferred by"), _("Picking Wave"), wave.id, wave.name)
-                picking.message_post(body=message_body)
-                picking_todo.add(picking.id)
-        if picking_todo:
-            self.pool.get('stock.picking').action_done(cr, uid, list(picking_todo), context=context)
-        return self.write(cr, uid, ids, {'state': 'done'}, context=context)
+    @api.multi
+    def done(self):
+        pickings_todo = self.env['stock.picking']
+        for picking in self.mapped('picking_ids').filtered(lambda x: x.state not in ('cancel', 'done')):
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py]: Do not use 'x', use a meaningfull variable naem like 'picking'.

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 12]
```diff
@@ -1,77 +1,66 @@
...
+            if picking.state != 'assigned':
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py]: I would move the check outside of the main operational loop, using an if (any (picking.state ... )).

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 13]
```diff
@@ -1,77 +1,66 @@
...
+                raise UserError(_('Some pickings are still waiting for goods. Please check or force their availability before setting this wave to done.'))
+            message_body = "<b>%s:</b> %s <a href=#id=%s&view_type=form&model=stock.picking.wave>%s</a>" % (_("Transferred by"), _("Picking Wave"), picking.wave_id.id, picking.wave_id.name)
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py]: Do not use an intermediate variable used only once.

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 14]
```diff
@@ -1,77 +1,66 @@
...
+            picking.message_post(body=message_body)
+            pickings_todo += picking
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py]: You may use |= that avoid duplicates in the record set.

File: addons/stock_picking_wave/models/stock_picking_wave.py [hunk 15]
```diff
@@ -1,77 +1,66 @@
...
+        pickings_todo.action_done()
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/models/stock_picking_wave.py]: Just to avoid issues, do: if pickings_todo: pickings_todo.action_done .

File: addons/stock_picking_wave/views/stock_picking_wave_views.xml [hunk 16]
```diff
@@ -1,6 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
 <odoo>
         <record id="view_picking_wave_form" model="ir.ui.view">
-            <field name="name">picking.wave.form</field>
+            <field name="name">stock.picking.wave.form</field>
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/views/stock_picking_wave_views.xml]: Unnecessary diff.

File: addons/stock_picking_wave/views/stock_picking_wave_views.xml [hunk 17]
```diff
@@ -47,7 +48,7 @@
 
 
         <record id="view_picking_wave_tree" model="ir.ui.view">
-            <field name="name">picking.wave.tree</field>
+            <field name="name">stock.picking.wave.tree</field>
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/views/stock_picking_wave_views.xml]: Unnecessary diff.

File: addons/stock_picking_wave/wizard/__init__.py [hunk 18]
```diff
@@ -1 +1,4 @@
-import picking_to_wave
\ No newline at end of file
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+import picking_to_wave
```
Comment (tde-banana-odoo) [inline addons/stock_picking_wave/wizard/__init__.py]: Rename the file to stock_picking_to_wave.py (do it in the move commit).
